state/azure/blobstorage/blobstorage.go

```diff
@@ -210,107 +159,124 @@ func (r *StateStore) readFile(ctx context.Context, req *state.GetRequest) (*stat
 		return &state.GetResponse{}, err
 	}
 
-	bodyStream := resp.Body(azblob.RetryReaderOptions{})
-	data, err := io.ReadAll(bodyStream)
+	reader := blobDownloadResponse.Body
+	defer reader.Close()
+	blobData, err := io.ReadAll(reader)
 	if err != nil {
-		r.logger.Debugf("read file %s, err %s", req.Key, err)
-		return &state.GetResponse{}, err
+		return &state.GetResponse{}, fmt.Errorf("error reading az blob: %w", err)
+	}
+	err = reader.Close()
+	if err != nil {
+		return &state.GetResponse{}, fmt.Errorf("error closing az blob reader: %w", err)
 	}
 
-	contentType := resp.ContentType()
+	contentType := blobDownloadResponse.ContentType
 
 	return &state.GetResponse{
-		Data:        data,
-		ETag:        ptr.Of(string(resp.ETag())),
-		ContentType: &contentType,
+		Data:        blobData,
+		ETag:        ptr.Of(string(*blobDownloadResponse.ETag)),
+		ContentType: contentType,
 	}, nil
 }
 
 func (r *StateStore) writeFile(ctx context.Context, req *state.SetRequest) error {
-	accessConditions := azblob.BlobAccessConditions{}
+	modifiedAccessConditions := blob.ModifiedAccessConditions{}
 
 	if req.ETag != nil && *req.ETag != "" {
-		accessConditions.IfMatch = azblob.ETag(*req.ETag)
+		modifiedAccessConditions.IfMatch = ptr.Of(azcore.ETag(*req.ETag))
 	}
 	if req.Options.Concurrency == state.FirstWrite && (req.ETag == nil || *req.ETag == "") {
-		accessConditions.IfNoneMatch = azblob.ETag("*")
+		modifiedAccessConditions.IfNoneMatch = ptr.Of(azcore.ETagAny)
 	}
 
-	blobURL := r.containerURL.NewBlockBlobURL(getFileName(req.Key))
+	accessConditions := blob.AccessConditions{
+		ModifiedAccessConditions: &modifiedAccessConditions,
+	}
 
 	blobHTTPHeaders, err := r.createBlobHTTPHeadersFromRequest(req)
 	if err != nil {
 		return err
 	}
-	_, err = azblob.UploadBufferToBlockBlob(ctx, r.marshal(req), blobURL, azblob.UploadToBlockBlobOptions{
-		Metadata:         req.Metadata,
-		AccessConditions: accessConditions,
-		BlobHTTPHeaders:  blobHTTPHeaders,
-	})
-	if err != nil {
-		r.logger.Debugf("write file %s, err %s", req.Key, err)
 
+	uploadOptions := azblob.UploadBufferOptions{
+		AccessConditions: &accessConditions,
+		Metadata:         storageinternal.SanitizeMetadata(r.logger, req.Metadata),
+		HTTPHeaders:      &blobHTTPHeaders,
+		Concurrency:      16,
```

Comment:
Still think this is a lot. Means having 16 goroutines uploading the file. (The default is 1, so no parallel)

Reply:
Removing it so it can use the default